### PR TITLE
Fix softmax_lse layout to avoid padding into max_seqlen_q

### PR DESCRIPTION
**Pull request by GD06**
As we observed in our long sequence training, the padding of various sequence length could consume a lot of memory (several to tens GB per GPU). In this PR, we try to fix the issue by making softmax_lse a compact memory layout from {batch_size, num_heads, max_seqlen_q} to {num_heads, total_q}.

More details can be found in https://github.com/Dao-AILab/flash-attention/issues/725

### CODE REVIEW

File: csrc/flash_attn/flash_api.cpp [hunk 1]
```diff
@@ -524,7 +526,7 @@ mha_varlen_fwd(const at::Tensor &q,  // total_q x num_heads x head_size, total_q
 
     auto opts = q.options();
 
-    auto softmax_lse = torch::empty({batch_size, num_heads, max_seqlen_q}, opts.dtype(at::kFloat));
+    auto softmax_lse = torch::empty({num_heads, total_q}, opts.dtype(at::kFloat));
```
Comment (aaronenyeshi) [inline csrc/flash_attn/flash_api.cpp:529]: The arguments for device be passed in to allocate on device?
```suggestion
    auto softmax_lse = torch::empty({num_heads, total_q}, dtype=opts.dtype(at::kFloat), device=q.get_device());
```

File: csrc/flash_attn/flash_api.cpp [hunk 2]
```diff
@@ -947,7 +951,7 @@ mha_varlen_bwd(const at::Tensor &dout,  // total_q x num_heads, x head_size
     at::cuda::CUDAGuard device_guard{(char)q.get_device()};
 
     auto opts = q.options();
-    auto softmax_d = torch::empty({batch_size, num_heads, seqlen_q_rounded}, opts.dtype(at::kFloat));
+    auto softmax_d = torch::empty({num_heads, total_q + 128 * batch_size}, opts.dtype(at::kFloat));
```
Comment (aaronenyeshi) [inline csrc/flash_attn/flash_api.cpp:954]: same here, are we expecting this tensor to be on the device?
```suggestion
    auto softmax_d = torch::empty({num_heads, total_q + 128 * batch_size}, dtype=opts.dtype(at::kFloat), device=q.get_device());
```

**Review by aaronenyeshi — COMMENTED** [review]
qq - do we expect the softmax to be on the device? If so, then it should allocate via CUDA caching allocator and show up in memory snapshots.

**Comment by GD06** [comment]
> qq - do we expect the softmax to be on the device? If so, then it should allocate via CUDA caching allocator and show up in memory snapshots.

Yes, we expect softmax_lse to be on the device. But I thought `opts = q.options()` already cloned the device info.

**Comment by jianyuh** [comment]
For the general supports for non var seq cases, to save memory for softmax_lse (https://github.com/GD06/flash-attention/pull/4)

- Install

```
with-proxy python setup.py develop 2>&1 | tee out.log
```
- Test
```
 pytest  tests/test_flash_attn.py -k test_flash_attn_causal -x 2>&1 | tee log.txt
 pytest  tests/test_flash_attn.py -k test_flash_attn_bwd_transpose -x 2>&1 | tee log.txt
```


https://www.internalfb.com/intern/paste/P1205198147/

 pytest  tests/test_flash_attn.py -x 2>&1 | tee log.txt
P1219468821

```
........................................................................ [ 99%]
........................................................................ [ 99%]
.....................................................................    [100%]

============== 201764 passed, 84480 skipped in 2091.37s (0:34:51) ==============
```
